### PR TITLE
Move the Security API to Mutiny

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by gsmet** [comment]
Looks like CI is consistently unhappy :).

**Comment by sberyozkin** [comment]
Wow, I can finally learn few things about Mutiny :-)

File: bom/runtime/pom.xml [hunk 1]
```diff
@@ -182,7 +182,7 @@
         <mockito.version>3.3.3</mockito.version>
         <jna.version>5.3.1</jna.version>
         <antlr.version>4.7.2</antlr.version>
-        <quarkus-security.version>1.0.1.Final</quarkus-security.version>
+        <quarkus-security.version>1.1.0.Beta1</quarkus-security.version>
```
Comment (sberyozkin) [inline bom/runtime/pom.xml:185]: Hi Stuart, it is not really related to this PR, would it make sense to move `quarkus-security` into `quarkus` eventually ? I've added a couple of exceptions into `quarkus-vertx-http` which should probably be in `quarkus-security`, and we can tweak things faster there

Comment (sberyozkin) [reply bom/runtime/pom.xml:185]: by the way, I'm not asking to move it now :-), only a question/suggestion as I recall you were contemplating it at some point

**Comment by sberyozkin** [comment]
@stuartwdouglas Hi Stuart I've looked through all the PR and it looks fine, I guess somewhere the UNI replacement may need to be tweaked a bit for the tests to pass, otherwise LGTM

**Comment by stuartwdouglas** [comment]
CI is happy now, one of the integration tests was missing a dependency on quarkus-resteasy-mutiny

**Comment by sberyozkin** [comment]
@stuartwdouglas Hi, thanks, I can't see anything wrong really with the PR even though it is the first time I'm looking closely at the Mutiny code :-), so I'll be happy to approve.

**Review by sberyozkin — APPROVED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/QuarkusErrorHandler.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ public void handle(RoutingContext event) {
                     public void run() {
                         event.response().end();
                     }
-                });
+                }).subscribeAsCompletionStage();
```
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/QuarkusErrorHandler.java:56]: Not totally sure what you are doing here.
The subscription will trigger the computation.
But you may want to be sure that an error has not been emitted.

**Review by cescoffier — CHANGES_REQUESTED** [review]
The new version of Mutiny provides a better way to create Uni with `null`. That would replace a bunch of empty optionals.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java [hunk 3]
```diff
@@ -48,15 +49,15 @@
             return context.runBlocking(new Supplier<SecurityIdentity>() {
                 @Override
                 public SecurityIdentity get() {
-                    return authenticate(request, vertxContext).join();
+                    return authenticate(request, vertxContext).await().indefinitely();
```
[the remark this answers is not in the document]
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:52]: @cescoffier the expectation here has been that Vertx client will not block itself if establishing a connection to OIDC is required. Can it block ?

**Comment by stuartwdouglas** [comment]
I have updated as per @cescoffier 's comments.

File: extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java [hunk 4]
```diff
@@ -74,39 +77,42 @@ public boolean isAnonymous() {
...
+            @Override
+            public Boolean apply(List<?> o) {
+                Boolean result = null;
+                //if any are true we return true
+                //otherwise if all are null we return null
+                //if some are false and some null we return false
+                for (Object i : o) {
+                    if (i != null) {
+                        boolean val = (boolean) i;
+                        if (val) {
```
Comment (cescoffier) [inline extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:107]: `return val`?

**Review by cescoffier — APPROVED** [review]
Just a very small simplification that could be done.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:52]: this code waits for an item, even if the underlying communication and processing in non-blocking, this code is waiting for the result.

**Comment by gsmet** [comment]
I rebased and force pushed. I don't think Clément's comment is blocking so I'll merge as soon as CI is green.

**Comment by gsmet** [comment]
The JDK 8 build failed in the AMQP tests with:
```
[ERROR] testCodeUpdate  Time elapsed: 5.585 s  <<< ERROR!
java.lang.IllegalStateException: Arc context propagation was attempted but the container is not running.
    at io.quarkus.arc.runtime.context.ArcContextProvider.lambda$currentContext$2(ArcContextProvider.java:35)
    at io.smallrye.context.impl.ActiveContextState.<init>(ActiveContextState.java:18)
    at io.smallrye.context.impl.CapturedContextState.begin(CapturedContextState.java:35)
    at io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:213)
    at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.onNext(ContextPropagatorOnFlowableCreateAction.java:50)
    at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drainLoop(FlowableFlatMap.java:506)
    at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drain(FlowableFlatMap.java:371)
    at io.reactivex.internal.operators.flowable.FlowableFlatMap$InnerSubscriber.onNext(FlowableFlatMap.java:670)
    at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.lambda$onNext$2(ContextPropagatorOnFlowableCreateAction.java:50)
    at io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:215)
    at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.onNext(ContextPropagatorOnFlowableCreateAction.java:50)
    at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:117)
    at io.reactivex.processors.AsyncProcessor.subscribeActual(AsyncProcessor.java:242)
    at io.reactivex.Flowable.subscribe(Flowable.java:14935)
    at io.reactivex.Flowable.subscribe(Flowable.java:14882)
    at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onNext(FlowableFlatMap.java:163)
    at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.lambda$onNext$2(ContextPropagatorOnFlowableCreateAction.java:50)
    at io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:215)
    at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.onNext(ContextPropagatorOnFlowableCreateAction.java:50)
    at io.reactivex.internal.util.HalfSerializer.onNext(HalfSerializer.java:45)
```
there must be a race condition somewhere.

/cc @FroMage @cescoffier 

Not blocking for this PR though.

**Comment by FroMage** [comment]
I'm not sure what that has to do with AMQP, or why they use rxjava2, or why the container stopped running.

**Pull request merged**